002-es-modules.md

```diff
@@ -0,0 +1,511 @@
+/opt/local/non-local-deps/node_modules -> $PREFIX/lib/node (etc.)
+```
+
+And nest as many times as needed.
+
+#### 4.3.3. Removal of fallthrough behavior.
+
+Exact algorithm TBD.
+
+#### 4.3.4. Errors from new path behavior.
+
+In the case that an `import` statement is unable to find a module, Node should
+make a **best effort** to see if `require` would have found the module and
+print out where it was found, if `NODE_PATH` was used, if `HOME` was used, etc.
+
+#### 4.4. Shipping both ESM and CJS
+
+When a `package.json` main is encountered, file extension searches are used to
+provide a means to ship both ESM and CJS variants of packages. If we have two
+entry points `index.mjs` and `index.js` setting `"main":"./index"` in
+`package.json` will make Node pick up either, depending on what is supported.
+
+##### 4.4.1. Excluding main
+
+Since `main` in `package.json` is entirely optional even inside of npm
+packages, some people may prefer to exclude main entirely in the case of using
+`./index` as that is still in the Node module search algorithm.
+
+### 4.5. ESM Evaluation
+
+#### 4.5.1. Environment Variables
+
+ESM will not be bootstrapped with magic variables and will await upcoming specifications in order to provide such behaviors in a standard way. As such, the following variables are changed:
+
+| Variable | Exists | Value |
+| ---- | ---- | --- |
+| this | y | [`undefined`](https://tc39.github.io/ecma262/#sec-module-environment-records-getthisbinding) |
+| arguments | n | |
+| require | n | |
+| module | n | |
+| exports | n | |
+| __filename | n | |
+| __dirname | n | |
+
+Like normal scoping rules, if a variable does not exist in a scope, the outer scope is used to find the variable. Since ESM are always strict, errors may be thrown upon trying to use variables that do not exist globally when using ESM.
+
+##### 4.5.1.1. Standardization Effort
+
+[Efforts are ongoing to reserve a specifier](https://github.com/whatwg/html/issues/1013) that will be compatible in both Browsers and Node. Tentatively it will be `js:context` and export a single `{url}` value.
+
+##### 4.5.1.2. Getting CJS Variables Workaround
+
+Although heavily advised against, you can have a CJS module sibling for your ESM that can export these things:
+
+```js
+// expose.js
+module.exports = {__dirname};
+```
+
+```js
+// use.mjs
+import expose from './expose.js';
+const {__dirname} = expose;
+```
+
+### 4.6. ES consuming CommonJS
+
+After *any* CJS finishes evaluation, it will be placed into the same cache as ESM.
+The value of what is placed in the cache will reflect a single default export pointing to the value of `module.exports` at the time evaluation ended.
+
+Essentially after any CJS completes evaluation:
+
+1. if there was an error, place the error in the ESM cache and return
+2. let `export` be the value of `module.exports`
+3. if there was an error, place the error in the ESM cache and return
+4. create an ESM with `{default:module.exports}` as its namespace
```

Comment:
So it's as if it was wrapped with a module that did `export default require('CJS')`